downloader/src/main/kotlin/VcsHost.kt

```diff
@@ -0,0 +1,172 @@
+/*
+ * Copyright (C) 2019 Bosch Software Innovations GmbH
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ * License-Filename: LICENSE
+ */
+
+package com.here.ort.downloader
+
+import com.here.ort.model.VcsInfo
+import com.here.ort.model.VcsType
+import com.here.ort.utils.hasFragmentRevision
+
+import java.net.URI
+import java.net.URISyntaxException
+import java.nio.file.Paths
+
+import kotlin.reflect.full.primaryConstructor
+
+/**
+ * A class to handle VCS-host-specific information.
+ */
+sealed class VcsHost(val url: String) {
+    companion object {
+        private val ALL = VcsHost::class.sealedSubclasses
+
+        /**
+         * Get the VCS host that is applicable for [url].
+         */
+        fun fromUrl(url: String): VcsHost? =
+            ALL.asSequence().map {
+                it.primaryConstructor!!.call(url)
+            }.find {
+                it.isApplicable()
+            }
+    }
+
+    /**
+     * The [URI] for [url] if it can be parsed, or null otherwise.
+     */
+    protected val uri = try {
+        URI(url)
+    } catch (e: URISyntaxException) {
+        null
+    }
+
+    /**
+     * The base hostname of VCS host.
+     */
+    abstract val hostname: String
+
+    /**
```

Comment:
Would it make sense to say a ```VcsHost``` can support a set of ```VcsType``` and represent that as
```fun getSupportedVcsTypes(): Set<VcsType>```? That'd require refactoring ```VcsType``` to enum though.

Reply:
Well, when using a `Set` (not an `EnumSet`) no refactoring of `VcsType` to an `enum` is necessary. Will do that!